src/libstd/sys/wasi/fs.rs

```diff
@@ -616,14 +616,37 @@ fn open_at(fd: &WasiFd, path: &Path, opts: &OpenOptions) -> io::Result<File> {
     Ok(File { fd })
 }
 
+/// Get pre-opened file descriptors and their paths.
+fn get_paths() -> Result<Vec<(PathBuf, WasiFd)>, wasi::Error> {
+    use crate::sys::os_str::Buf;
+
+    let mut paths = Vec::new();
+    for fd in 3.. {
+        match wasi::fd_prestat_get(fd) {
+            Ok(wasi::Prestat { pr_type: wasi::PREOPENTYPE_DIR, u }) => unsafe {
+                let name_len = u.dir.pr_name_len;
+                let mut buf = vec![0; name_len];
+                wasi::fd_prestat_dir_name(fd, &mut buf)?;
+                let buf = Buf { inner: buf };
+                let path = PathBuf::from(OsString { inner: buf });
+                paths.push((path, WasiFd::from_raw(fd)));
+            },
+            Ok(_) => (),
+            Err(wasi::EBADF) => break,
+            Err(err) => return Err(err),
+        }
+    }
+    paths.shrink_to_fit();
+    Ok(paths)
```

Comment:
Since the Vec can not/should not be changed, will it be more appropriate to return a boxed slice here? (`paths.into_boxed_slice()`)

Reply:
Yes, good catch.